PYIC 4961 - Automatic Removal of Weaker Operational Profiles

sammyiyke opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

- Adds automatic removal of weaker operational profiles during new IPV journey
- Adds automatic removal of weaker operational profiles upon return from F2F
- Adds tests for automatic removal of weaker operational profiles during new IPV journey
- Adds tests for automatic removal of weaker operational profiles upon return from F2F
- Fixes bug with cases where the evidence is empty for a VC

<!-- Describe the changes in detail - the "what"-->

### Why did it change

[PYIC 4255 - Automated Removal of Weaker Operational Profiles in VC Store upon Awarding Superior GPG45 Profiles](https://govukverify.atlassian.net/browse/PYIC-4255)

To ensure that CORE maintains the most relevant and highest strength identity profiles for users, thereby enhancing the security and integrity of the identity verification process.

<!-- Describe the reason these changes were made - the "why" -->

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-4255](https://govukverify.atlassian.net/browse/PYIC-4255)


[PYIC-4255]: https://govukverify.atlassian.net/browse/PYIC-4255?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

RogVanGDS committed:
Revert "Revert "PYIC-4563: Added journey-map routing for M2B KBV fail and dropout pages""

This reverts commit f107e031

Wynndow committed:
PYIC-4982: Defer setting of signing key ID

The process-cri-callback lambda uses the CriApiService to generated
token requests for CRIs.

These token requests are signed, and we use a custom KMS signer to do
this. The signer needs a KMS key Id configuring on it.

We were doing this in the initialisation phase of the lambda. Moving to
snapstart would meant that value was permanently baked in to the lambda.
This would mean we couldn't switch signing keys easily if required.

This change defers the setting of the key id to the runtime. We use a
KMS signer when running in a live enviroment, and a normal ES245 signer
in the tests, which is why we need to check if the signer is of a
particular type before trying to set the KMS key ID on it.

sammyiyke committed:
PYIC-4255: Refactor - Renames VC_HMRC_MIGRATION_WITH_NO_EVIDENCE to VC_INHERITED_IDENTITY_MIGRATION_WITH_NO_EVIDENCE for broader compatibility

sammyiyke committed:
PYIC-4255: Refactor - Renames VC_HMRC_MIGRATION_WITH_NO_EVIDENCE to VC_INHERITED_IDENTITY_MIGRATION_WITH_NO_EVIDENCE for broader compatibility

sammyiyke committed:
PYIC-4255: Adds Automatic Removal of Weaker Operational Profiles - New IPV Journey and Tests: Automatic Removal of Weaker Operational Profiles - New IPV Journey

sammyiyke committed:
PYIC-4255: Adds Automatic Removal of Weaker Operational Profiles - New IPV Journey and Tests: Automatic Removal of Weaker Operational Profiles - New IPV Journey

Wynndow committed:
PYIC-4979: Switch build to use x86_64

This is the first step in enabling snapstart in the build env. You can't
change architecture and enable snapstart at the same time, so this is
step one of a two step process.

sammyiyke committed:
PYIC-4255: Adds Automatic Removal of Weaker Operational Profiles - F2F Journey

sammyiyke committed:
PYIC-4255: Adds Automatic Removal of Weaker Operational Profiles - F2F Journey

Joe-Edwards-GDS committed:
BAU: Update GH pages actions

sammyiyke committed:
PYIC-4255: Adds Tests for Automatic Removal of Weaker Operational Profiles - F2F Journey

sammyiyke committed:
PYIC-4255: Adds Tests for Automatic Removal of Weaker Operational Profiles - F2F Journey

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -451,11 +452,26 @@ private boolean achievedWithGpg45Profile(
+            if (!VcHelper.isOperationalProfileVc(credential)) {
+                gpg45Credentials.add(credential);
+            }
+        }
```
We've got a helper for this already - `VcHelper.filterVCBasedOnProfileType`

sammyiyke replied:
True, but this takes `VcStoreItem` instead of the `SignedJWT` we are trying to filter at this stage. 

This would mean creating another instance of `VcHelper.filterVCBasedOnProfileType` to handle `SignedJWT`.

Joe-Edwards-GDS replied:
Ugh, yeah the difference between `SignedJWT` and `VCStoreItem` is awkward. I'm hoping we can fix this as part of PYIC-3908, but we might have to do some groundwork first.

If you can't see a neat fix then perhaps this is ok for now.

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -451,11 +452,26 @@ private boolean achievedWithGpg45Profile(
+                for (SignedJWT credential : credentials) {
+                    if (VcHelper.isOperationalProfileVc(credential)) {
+                        verifiableCredentialService.deleteVcStoreItem(
+                                auditEventUser.getUserId(), HMRC_MIGRATION_CRI);
```
This is a bit weird - when checking the VC, we're inspecting the VC and using the presence of the `vot` field, but when deleting we're hardcoding `HMRC_MIGRATION_CRI`.

Perhaps the easiest would be to iterate through `vcStoreItems` and delete the `HMRC_MIGRATION_CRI` entry if there is one?

(More generally I think we need to start being more consistent about whether we're using 'presence of VOT' or 'issuer = HMRC Migration' to identify operational profiles - at the moment they're equivalent, but they may not be in future)

Joe-Edwards-GDS commented on lambdas/evaluate-gpg45-scores/src/main/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandler.java:
```diff
@@ -129,13 +132,33 @@ public Map<String, Object> handleRequest(JourneyRequest event, Context context)
+                if (!VcHelper.isOperationalProfileVc(credential)) {
+                    gpg45Credentials.add(credential);
+                }
+            }
```
(ditto with the helper)

Joe-Edwards-GDS replied:
In this case I think we could filter the VC list _before_ parsing into `SignedJWT`s? - I don't think there's any need to handle non-GPG45 VCs here.

sammyiyke replied:
I don't think `VcHelper.filterVCBasedOnProfileType` filters operational profiles. It currently returns either `OPERATIONAL_CRIS` or none `OPERATIONAL_CRIS`. It does not filter out the operational profiles matching `HMRC_MIGRATION_CRI`.

Joe-Edwards-GDS replied:
Yeah, this is what I meant in the other comment... At the moment we have two different ways of identifying operational profiles:
- A VC with a `vot` matching an operational profile
- A VC from `HMRC_MIGRATION_CRI`

At the moment these are exactly equivalent, but it'd be nice to be consistent (or be clear about when we should use one vs. the other).

In this case, could you do `VcHelper.filterVCBasedOnProfileType(vcStoreItems, ProfileType.GPG45)`? I think that better reflects what you're trying to do - i.e. find all the GPG45 VCs.

Joe-Edwards-GDS commented on lambdas/evaluate-gpg45-scores/src/main/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandler.java:
```diff
@@ -129,13 +132,33 @@ public Map<String, Object> handleRequest(JourneyRequest event, Context context)
+                            ipAddress);
+
+            if (configService.enabled(CoreFeatureFlag.INHERITED_IDENTITY)) {
+                if (checkForMatchingGpg45Profile.equals(JOURNEY_MET)) {
```
This condition is a bit weird - I'd prefer to apply this 'directly' when matching a profile (rather than checking the journey response).

Can we make that method return the matched profile (or just true/false) and move the routing-specific logic here? - Then this check can be `if (metProfile) { ... }` or similar

Joe-Edwards-GDS commented on lambdas/evaluate-gpg45-scores/src/main/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandler.java:
```diff
@@ -129,13 +132,33 @@ public Map<String, Object> handleRequest(JourneyRequest event, Context context)
+                        if (VcHelper.isOperationalProfileVc(credential)) {
+                            verifiableCredentialService.deleteVcStoreItem(
+                                    userId, HMRC_MIGRATION_CRI);
+                        }
```
As above

Joe-Edwards-GDS reviewed (commented):
A couple of comments for you to think about :)

Joe-Edwards-GDS committed:
Merge pull request #1641 from govuk-one-login/bau-update-gh-pages-steps

BAU: Update GH pages actions

DanCorderIPV committed:
BAU: Update DCMAW tests with feedback from mobile team. Also move pact test helper code to it's own library.

Wynndow committed:
Merge pull request #1638 from govuk-one-login/pyic-4979-switch-build-to-intel

PYIC-4979: Switch build to use x86_64

RogVanGDS committed:
Merge pull request #1643 from govuk-one-login/PYIC-4563-fix

Revert "Revert "PYIC-4563: Added journey-map routing for M2B KBV fail…

DanCorderIPV committed:
Merge commit 'f52648eb406af85b3a7f85452795cf3b82689294' into BAU-mobile-contract-test-tweaks

# Conflicts:
#	.secrets.baseline

DanCorderIPV committed:
BAU: Suppress over zealous sonarcloud warning and fix spelling

Wynndow committed:
PYIC-4982: Create KMS signer with factory

This refactors the creation of KMS signer to use a factory. This makes
it much more obvious that a signer needs the key ID setting on it before
use.

Wynndow committed:
Merge pull request #1637 from govuk-one-login/pyic-4982-check-lambdas-for-snapstart-readiness

PYIC-4982: Defer setting of signing key ID

Wynndow committed:
PYIC-4980: Activate snapstart in build and prep staging

This turns on snapstart in the build env, and preps staging for it. We
can probably just turn snapstart on in staging immediately. We'll
potentially want to to int/prod together once we're comfortable with it
in the lower envs.

DanCorderIPV committed:
BAU: Add tests for new library

DanCorderIPV committed:
Merge commit '87b2e4f1f0b8266a99cf6608b77aeb4b60a3463f' into BAU-mobile-contract-test-tweaks

# Conflicts:
#	.secrets.baseline
#	lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/addressCri/ContractTest.java
#	lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/bavCri/ContractTest.java
#	lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/cicCri/ContractTest.java
#	lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/dcmawCri/ContractTest.java
#	lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/drivingLicenceCri/ContractTest.java
#	lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/hmrcKbvCri/ContractTest.java
#	lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/ninoCri/ContractTest.java
#	lambdas/process-cri-callback/src/test/java/uk/gov/di/ipv/core/processcricallback/pact/passportCri/CredentialTests.java

DanCorderIPV committed:
BAU: Spotless and secrets

Wynndow committed:
Merge pull request #1654 from govuk-one-login/pyic-4980-enable-snapstart-in-build

PYIC-4980: Activate snapstart in build and prep staging

Wynndow committed:
PYIC-4980: Remove provisioned concurrency in build

Snapstart and provisioned concurrency are not compatible. Trying to use
both will not deploy.

Wynndow committed:
Merge pull request #1656 from govuk-one-login/pyic-4980-disable-provisioned-concurrency-in-build

PYIC-4980: Remove provisioned concurrency in build

DanCorderIPV committed:
BAU: Add test report generation

DanCorderIPV committed:
Merge commit '6d223ef9cfafa4757ddae28e6f612b72babe083c' into BAU-mobile-contract-test-tweaks

# Conflicts:
#	.secrets.baseline

DanCorderIPV committed:
Merge pull request #1651 from govuk-one-login/BAU-mobile-contract-test-tweaks

BAU: Update DCMAW tests with feedback from mobile team

Wynndow committed:
PYIC-4980: Disable snapstart in build temporarily

Snapstart needs provisioned concurrency to be disabled in order to
activate. And provisioned concurrency needs snapstart to not be active
to be disabled.

This temporarliy turns snapstart off in build, so we can stop the
provisioned concurrency. Once that has happened we can turn snapstart
back on.

Once build is running, I'll be experimenting with my dev env to try and
find a way to do this in prod without having to run without provisioned
concurrency or snapstart.

Wynndow committed:
Merge pull request #1657 from govuk-one-login/pyic-4980-disable-snapstart-in-build

PYIC-4980: Disable snapstart in build temporarily

Wynndow committed:
PYIC-4980: Re-enable snapstart in build

Hopefully now we don't have provisioned concurrency running in build,
and so we can enable snapstart.

DanCorderIPV committed:
BAU: Update redirect URL to work around pact bug and update auth code to be a valid UUID

DanCorderIPV committed:
Merge pull request #1659 from govuk-one-login/BAU-more-mobile-contract-test-tweaks

BAU: DCMAW pact test tweaks

Wynndow committed:
Merge pull request #1658 from govuk-one-login/pyic-4980-reenable-snapstart-in-build

PYIC-4980: Re-enable snapstart in build

sammyiyke committed:
PYIC-4255: Implements changes from PR reviews

sammyiyke committed:
PYIC-4255: Implements changes from PR reviews

sammyiyke committed:
PYIC-4255: Fixed imports class

sammyiyke committed:
PYIC-4255: Fixed imports class

sammyiyke committed:
PYIC-4255: Fixed imports class

sammyiyke committed:
PYIC-4255: Fixed imports class

sammyiyke committed:
PYIC-4255: Fixed failing tests

sammyiyke committed:
Merge remote-tracking branch 'origin/PYIC-4961-Automatic-Removal-of-Weaker-Operational-Profiles-New-IPV-Journey' into PYIC-4961-Automatic-Removal-of-Weaker-Operational-Profiles-New-IPV-Journey

# Conflicts:
#	lambdas/check-existing-identity/src/test/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandlerTest.java

sammyiyke committed:
PYIC-4255: Fixes failing sonar code smells

sammyiyke committed:
PYIC-4255: Fixes failing sonar code smells
